src/gcl/MeshVelocityEdgeAlg.C

```diff
@@ -0,0 +1,183 @@
+      const int* lrscv = meSCS_->adjacentNodes();
+      const int* scsIpEdgeMap = meSCS_->scsIpEdgeOrd();
+
+      auto& scrView = edata.simdScrView;
+
+      const auto& mCoords = scrView.get_scratch_view_2D(modelCoordsID);
+      const auto& dispNp1 = scrView.get_scratch_view_2D(meshDispNp1ID);
+      const auto& dispN = scrView.get_scratch_view_2D(meshDispNID);
+
+      DoubleType scs_coords_n[19][AlgTraits::nDim_];
+      DoubleType scs_coords_np1[19][AlgTraits::nDim_];
+
+      for (int i = 0; i < 19; i++) {
+        for (int j = 0; j < AlgTraits::nDim_; j++) {
+          scs_coords_n[i][j] = 0.0;
+          scs_coords_np1[i][j] = 0.0;
+        }
+        for (int k = 0; k < AlgTraits::nodesPerElement_; k++) {
+          const DoubleType r =
+            isoCoordsShapeFcn_[i * AlgTraits::nodesPerElement_ + k];
+          for (int j = 0; j < AlgTraits::nDim_; j++) {
+            scs_coords_n[i][j] += r * (mCoords(k, j) + dispN(k, j));
+            scs_coords_np1[i][j] += r * (mCoords(k, j) + dispNp1(k, j));
+          }
+        }
+      }
+
+      for (int ip = 0; ip < AlgTraits::numScsIp_; ++ip) {
+
+        const int na = scsFaceNodeMap_[ip][0];
+        const int nb = scsFaceNodeMap_[ip][1];
+        const int nc = scsFaceNodeMap_[ip][2];
+        const int nd = scsFaceNodeMap_[ip][3];
+
+        DoubleType scs_vol_coords[8][3];
+
+        for (int j = 0; j < AlgTraits::nDim_; j++) {
+          scs_vol_coords[0][j] = scs_coords_n[na][j];
+          scs_vol_coords[1][j] = scs_coords_n[nb][j];
+          scs_vol_coords[2][j] = scs_coords_n[nc][j];
+          scs_vol_coords[3][j] = scs_coords_n[nd][j];
+          scs_vol_coords[4][j] = scs_coords_np1[na][j];
+          scs_vol_coords[5][j] = scs_coords_np1[nb][j];
+          scs_vol_coords[6][j] = scs_coords_np1[nc][j];
+          scs_vol_coords[7][j] = scs_coords_np1[nd][j];
+        }
+
+        DoubleType tmp = hex_volume_grandy(scs_vol_coords);
+        DoubleType tmp2 = gamma1 * tmp / dt;
+
+        for (int si = 0; si < edata.numSimdElems; ++si) {
+          const auto edges = ngpMesh.get_edges(
+            stk::topology::ELEM_RANK,
+            ngpMesh.fast_mesh_index(edata.elemInfo[si].entity));
+
+          // Edge for this integration point
+          const int nedge = scsIpEdgeMap[ip];
+          // Index of "left" node in the element relations
+          const int iLn = lrscv[2 * ip];
+
+          // Nodes connected to this edge
+          const auto edgeID = ngpMesh.fast_mesh_index(edges[nedge]);
+          const auto edge_nodes =
+            ngpMesh.get_nodes(stk::topology::EDGE_RANK, edgeID);
+
+          // Left node comparison
+          const auto lnElemId = edata.elemInfo[si].entityNodes[iLn];
+          const auto lnEdgeId = edge_nodes[0];
+
+          const double sign = (lnElemId == lnEdgeId) ? 1.0 : -1.0;
+
+          Kokkos::atomic_add(
+            &edgeSweptVol.get(edgeID, 0), stk::simd::get_data(tmp, si) * sign);
+          Kokkos::atomic_add(
```

Comment:
There's an "ops" helper for doing these field update with simd types. see for instance the nodalgradpopen algorithm.